tools/transform.cpp

```diff
@@ -285,6 +285,71 @@ static expr preprocess(Transform &t, const set<expr> &qvars0,
 }
 
 
+static expr
+encode_undef_refinement_per_elem(
+    const Type &ty, const StateValue &sva, const expr &a2,
+    const expr &b, const expr &b2) {
+  const auto *aty = ty.getAsAggregateType();
+  if (!aty)
+    return (sva.non_poison && sva.value == a2).implies(b == b2);
+
+  StateValue sva2{expr(a2), expr()};
+  StateValue svb{expr(b), expr()}, svb2{expr(b2), expr()};
+  expr result = true;
+
+  for (unsigned i = 0; i < aty->numElementsConst(); ++i) {
+    if (aty->isPadding(i))
+      continue;
+
+    result &= encode_undef_refinement_per_elem(aty->getChild(i),
+        aty->extract(sva, i), aty->extract(sva2, i).value,
+        aty->extract(svb, i).value, aty->extract(svb2, i).value);
+  }
+  return result;
+}
+
+static expr
+encode_undef_refinement(const Type &type, const State::ValTy &ap,
+                        const State::ValTy &bp, set<expr> &qvars) {
+  // Undef refinement: (src-nonpoison /\ src-nonundef) -> tgt-nonundef
+  //
+  // Full refinement formula:
+  //   forall N_tgt, exists N_src,
+  //    (src_nonpoison /\ forall N_src', retval_src(N_src)==retval_src(N_src'))
+  //      -> forall N_tgt', retval_tgt(N_tgt) == retval_tgt(N_tgt')
+  //
+  // Since retval_tgt is the return value of the target program, it never uses
+  // N_src. Therefore, N_tgt' can be moved to the outer forall quantifier.
+  // Also, since bitvector isnt an empty set, N_src' can move to the
+  // existential quantifier too.
+  //
+  //   forall N_tgt, N_tgt', exists N_src, N_src',
+  //      (src_nonpoison /\ retval_src(N_src) == retval_src(N_src')) ->
+  //      retval_tgt(N_tgt) == retval_tgt(N_tgt')
+
+  if (dynamic_cast<const VoidType *>(&type))
+    return expr(true);
+
+  vector<pair<expr, expr>> repls_src, repls_tgt;
+  for (auto &u : ap.second) {
+    expr newvar = expr::mkFreshVar("undef", u);
+    repls_src.emplace_back(u, expr::mkFreshVar("undef", u));
```

Comment:
I think there's a typo here. it should use newvar, right?

Reply:
You're right, fixed